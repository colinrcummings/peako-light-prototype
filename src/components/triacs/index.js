export { default as TriacsChart } from "./TriacsChart";
export { default as TriacsLights } from "./TriacsLights";
