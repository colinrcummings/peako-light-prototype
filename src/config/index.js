/*
  re-exporting config for x second loop
*/
export { loopEnd, lights } from "./60";
