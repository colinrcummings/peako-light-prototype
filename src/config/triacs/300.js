import {
  LIGHT_RED_2,
  LIGHT_GREEN_2,
  LIGHT_BLUE,
  LIGHT_GREEN_1,
  LIGHT_RED_1,
  RED,
  GREEN,
  BLUE
} from "../../constants";

const fadeDurationMs = 2000;

export const tickMs = 200; // 2000 ms fade / 10 fade powers

export const tickCount = 300; // 60000 ms / 200 ms tick

const ticksPerSec = 1000 / tickMs;

const fadeFor = fadeDurationMs / tickMs;

export const lights = {
  // Red 2 comes on at 32 increases to full at 34, stays full 34-41, and fades off from 41-43;
  [LIGHT_RED_2]: {
    id: LIGHT_RED_2,
    color: RED,
    startFadeIn: 32 * ticksPerSec,
    startFadeOut: 41 * ticksPerSec,
    fadeFor
  },
  // Green 2 comes on at 20, increases to full at 22, stays full 22-36, and fades off from 36-38;
  [LIGHT_GREEN_2]: {
    id: LIGHT_GREEN_2,
    color: GREEN,
    startFadeIn: 20 * ticksPerSec,
    startFadeOut: 36 * ticksPerSec,
    fadeFor
  },
  // Blue is on continually;
  [LIGHT_BLUE]: {
    id: LIGHT_BLUE,
    color: BLUE,
    alwaysOn: true
  },
  // Green 1 comes on at 5, increases to full at 7, stays full 7-21, and fades off from 21-23;
  [LIGHT_GREEN_1]: {
    id: [LIGHT_GREEN_1],
    color: GREEN,
    startFadeIn: 5 * ticksPerSec,
    startFadeOut: 21 * ticksPerSec,
    fadeFor
  },
  // Red 1 comes on at 0, increases to full at 2, stays full 2-9, and fades off from 9-11;
  [LIGHT_RED_1]: {
    id: LIGHT_RED_1,
    color: RED,
    startFadeIn: 0 * ticksPerSec,
    startFadeOut: 9 * ticksPerSec,
    fadeFor
  }
};
