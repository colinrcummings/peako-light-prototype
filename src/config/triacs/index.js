/*
  available light powers determined with `rbddimmer_serial_monitor_dim.ino`
  see: https://github.com/RobotDynOfficial/RBDDimmer/issues/4#issuecomment-475065928
*/
export const LOW_POWER = 21;

export const HIGH_POWER = 50;

// ascending order (reverse for descending)
export const FADE_POWERS = [26, 28, 29, 30, 31, 32, 33, 34, 35, 36];

/*
  300 tick loop @ 200 ms per tick (fade)
*/
export * from "./300";
