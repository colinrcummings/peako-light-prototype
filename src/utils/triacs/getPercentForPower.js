import {
  LOW_POWER_POSITION,
  HIGH_POWER_POSITION,
  FADE_POWER_POSITIONS
} from "../../config/triacs";

const getPercentForPower = power => {
  const powerPositions = [
    LOW_POWER_POSITION,
    ...FADE_POWER_POSITIONS,
    HIGH_POWER_POSITION
  ];
  const numerator = powerPositions.indexOf(power);
  const denominator = powerPositions.length - 1;
  return numerator / denominator;
};

export default getPercentForPower;
