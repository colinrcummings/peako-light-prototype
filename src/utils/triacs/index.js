export { default as generatePowerPositionsForLight } from "./generatePowerPositionsForLight";
export { default as getBrightnessForPower } from "./getBrightnessForPower";
