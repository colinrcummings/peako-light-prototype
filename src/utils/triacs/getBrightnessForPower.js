import { LOW_POWER, HIGH_POWER, FADE_POWERS } from "../../config/triacs";

const getBrightnessForPower = power => {
  const powerPositions = [LOW_POWER, ...FADE_POWERS, HIGH_POWER];
  const numerator = powerPositions.indexOf(power);
  const denominator = powerPositions.length - 1;
  return numerator / denominator;
};

export default getBrightnessForPower;
