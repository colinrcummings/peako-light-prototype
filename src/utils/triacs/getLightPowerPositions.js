import { LOW_POWER, HIGH_POWER, FADE_POWERS } from "../../config/triacs";

const getLightPowerPositions = (lightConfig, tickCount) => {
  const { alwaysOn, startFadeIn, startFadeOut, fadeFor } = lightConfig;

  if (alwaysOn) {
    return new Array(tickCount).fill(HIGH_POWER);
  }

  const powerPositions = [];
  let position = 0;

  while (position < tickCount) {
    const lightIsFadingIn =
      position > startFadeIn && position <= startFadeIn + fadeFor;
    const lightIsFadingOut =
      position >= startFadeOut && position < startFadeOut + fadeFor;
    const lightIsOn =
      position > startFadeIn + fadeFor && position <= startFadeOut;

    if (lightIsFadingIn) {
      const fadeInPowerIdx = position - startFadeIn - 1;
      powerPositions.push(FADE_POWERS[fadeInPowerIdx]);
    } else if (lightIsFadingOut) {
      const fadeOutPowerIdx = position - startFadeOut;
      powerPositions.push([...FADE_POWERS].reverse()[fadeOutPowerIdx]);
    } else if (lightIsOn) {
      powerPositions.push(HIGH_POWER);
    } else {
      powerPositions.push(LOW_POWER);
    }

    position++;
  }

  return powerPositions;
};

export default getLightPowerPositions;
