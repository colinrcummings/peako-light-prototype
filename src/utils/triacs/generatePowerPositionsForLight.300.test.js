import { LOW_POWER, HIGH_POWER } from "../../config/triacs";
import { lights, tickCount, tickMs } from "../../config/triacs/300";
import generatePowerPositionsForLight from "./generatePowerPositionsForLight";

const getTicksOnLow = result =>
  result.filter(power => power === LOW_POWER).length;

const getTicksOnHigh = result =>
  result.filter(power => power === HIGH_POWER).length;

const ticksFor7Secs = 7000 / tickMs;
const ticksFor14Secs = 14000 / tickMs;
const ticksFor42Secs = 42000 / tickMs;
const ticksFor49Secs = 49000 / tickMs;
const ticksFor60Secs = 60000 / tickMs;

describe("generatePowerPositionsForLight for a tickCount of 300", () => {
  describe("for red2 light", () => {
    const { red2 } = lights;

    it("should return array of power positions with it on low for 49 seconds and on high for 7 seconds", () => {
      const result = generatePowerPositionsForLight(red2, tickCount);

      expect(result).toMatchSnapshot();
      expect(getTicksOnLow(result)).toBe(ticksFor49Secs);
      expect(getTicksOnHigh(result)).toBe(ticksFor7Secs);
    });
  });

  describe("for green2 light", () => {
    const { green2 } = lights;

    it("should return array of power positions with it on low for 42 seconds and on high for 14 seconds", () => {
      const result = generatePowerPositionsForLight(green2, tickCount);

      expect(result).toMatchSnapshot();
      expect(getTicksOnLow(result)).toBe(ticksFor42Secs);
      expect(getTicksOnHigh(result)).toBe(ticksFor14Secs);
    });
  });

  describe("for blue light", () => {
    const { blue } = lights;

    it("should return array of power positions with it on low for 0 seconds and on high for 60 seconds", () => {
      const result = generatePowerPositionsForLight(blue, tickCount);

      expect(getTicksOnLow(result)).toBe(0);
      expect(getTicksOnHigh(result)).toBe(ticksFor60Secs);
    });
  });

  describe("for green1 light", () => {
    const { green1 } = lights;

    it("should return array of power positions with it on low for 42 seconds and on high for 14 seconds", () => {
      const result = generatePowerPositionsForLight(green1, tickCount);

      expect(result).toMatchSnapshot();
      expect(getTicksOnLow(result)).toBe(ticksFor42Secs);
      expect(getTicksOnHigh(result)).toBe(ticksFor14Secs);
    });
  });

  describe("for red1 light", () => {
    const { red1 } = lights;

    it("should return array of power positions with it on low for 49 seconds and on high for 7 seconds", () => {
      const result = generatePowerPositionsForLight(red1, tickCount);

      expect(result).toMatchSnapshot();
      expect(getTicksOnLow(result)).toBe(ticksFor49Secs);
      expect(getTicksOnHigh(result)).toBe(ticksFor7Secs);
    });
  });
});
