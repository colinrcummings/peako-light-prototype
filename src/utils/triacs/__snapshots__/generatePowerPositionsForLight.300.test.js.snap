// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePowerPositionsForLight for a tickCount of 300 for green1 light should return array of power positions with it on low for 42 seconds and on high for 14 seconds 1`] = `
Array [
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  26,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  36,
  35,
  34,
  33,
  32,
  31,
  30,
  29,
  28,
  26,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
]
`;

exports[`generatePowerPositionsForLight for a tickCount of 300 for green2 light should return array of power positions with it on low for 42 seconds and on high for 14 seconds 1`] = `
Array [
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  26,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  36,
  35,
  34,
  33,
  32,
  31,
  30,
  29,
  28,
  26,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
]
`;

exports[`generatePowerPositionsForLight for a tickCount of 300 for red1 light should return array of power positions with it on low for 49 seconds and on high for 7 seconds 1`] = `
Array [
  20,
  26,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  36,
  35,
  34,
  33,
  32,
  31,
  30,
  29,
  28,
  26,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
]
`;

exports[`generatePowerPositionsForLight for a tickCount of 300 for red2 light should return array of power positions with it on low for 49 seconds and on high for 7 seconds 1`] = `
Array [
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  26,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  37,
  36,
  35,
  34,
  33,
  32,
  31,
  30,
  29,
  28,
  26,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
  20,
]
`;
